library/kani/src/invariant.rs

```diff
@@ -0,0 +1,66 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module introduces the `Invariant` trait as well as its implementation
+//! for primitive types.
+
+/// This trait should be used to specify and check type safety invariants for a
+/// type.
+///
+/// Note: Type safety might be checked automatically by Kani in the future.
+pub trait Invariant
+where
+    Self: Sized,
+{
+    fn is_safe(&self) -> bool;
+}
+
+impl Invariant for bool {
+    #[inline(always)]
+    fn is_safe(&self) -> bool {
+        let value = *self as u8;
+        value < 2
+    }
+}
+
+impl Invariant for char {
+    #[inline(always)]
+    fn is_safe(&self) -> bool {
+        let value = *self as u32;
+        value <= 0xD7FF || (0xE000..=0x10FFFF).contains(&value)
+    }
+}
+
+/// Any value is considered safe for the type
+macro_rules! trivial_invariant {
+    ( $type: ty ) => {
+        impl Invariant for $type {
+            #[inline(always)]
+            fn is_safe(&self) -> bool {
+                true
+            }
+        }
+    };
+}
```

Comment:
You don't need this. These are validity invariants and they should be enforced by the compiler.

Reply:
But won't this be an issue when we have automatically-derived implementations of `Invariant` (e.g., for a `struct`) which rely on member types to implement `Invariant`?

Reply:
Validity invariant has to be guaranteed at construction time, it's not something you can check after the fact.

Let's say we were to implement a safe `any_raw()` using the trait `Invariant`, it would look like this:
```rust
fn any_raw<T: Invariant>() -> T {
  // First ensure memory is initialized with a valid value.
  let maybe : MaybeUninit<T> = kani::any();
  kani::assume(kani::has_valid_value(maybe.as_ptr()));

  // Now ensure the object safety invariants are respected.
  let val = unsafe { maybe.assume_init() };
  kani::assume(val.is_safe());
  val
}
```